src/controllers/attemptController.js

```diff
@@ -0,0 +1,159 @@
+const { StatusCodes } = require("http-status-codes");
+const Attempt = require("../models/attemptModel");
+const { default: mongoose } = require("mongoose");
+const ObjectId = mongoose.Types.ObjectId;
+
+// add attempt
+const createAttempt = async (req, res) => {
+  try {
+    // find the latestAttempt
+    const latestAttempt = await Attempt.find()
+      .sort({ _id: -1 })
+      .findOne({ user: req.body.user })
+      .findOne({ test: req.body.test });
+    const newAttempt = new Attempt(req.body);
+    // increment
+    latestAttempt
+      ? (newAttempt.attemptNumber = latestAttempt.attemptNumber + 1)
+      : (newAttempt.attemptNumber = 1);
+    await newAttempt.save();
+    res.status(StatusCodes.CREATED).send({
+      message: "Attempt added successfully",
+    });
+  } catch (error) {
+    res.status(500).send({
+      message: error.message,
+    });
+  }
+};
+
+// get max score for every quiz by user, but for the latest attempt
+const maxScoreAttempts = async (userId) => {
+  try {
+    //rank
+    const attempts = await Attempt.aggregate([
+      {
+        $match: { user: userId },
+      },
+      {
+        $setWindowFields: {
+          partitionBy: "$test",
+          sortBy: { score: -1 },
+          //sortBy: { attemptNumber: -1 },
+          output: {
+            rankScoreForTest: {
+              $rank: {},
+            },
+          },
+        },
+      },
+      {
+        $match: { rankScoreForTest: 1 },
+      },
+      {
+        $project: { test: 1, maxScore: "$score" },
+      },
+    ]);
+    return attempts;
+    //
+  } catch (error) {
+    return error;
+  }
+};
+
+// get latest score for every quiz by user: sortBy attempNumber in descending order
+const latestScoreAttempts = async (userId) => {
```

Comment:
I found it a little difficult to understand which function in this file was the controller. I generally try to keep only controller functions in controller files and import any extra functions needed for the controller. What do you think of moving latestScoreAttempts and maxScoreAttempts to another file? Perhaps src/attempts.js?

Reply:
Moved latestScoreAttempts and maxScoreAttempts to src/functions/attempts.js